python_modules/dagster/dagster/core/execution/execution_result.py

```diff
@@ -0,0 +1,192 @@
+from abc import ABC, abstractmethod
+from typing import Any, Callable, List, Sequence, Union, cast
+
+import dagster._check as check
+from dagster.core.definitions import JobDefinition, NodeHandle
+from dagster.core.definitions.events import AssetMaterialization, AssetObservation, Materialization
+from dagster.core.definitions.utils import DEFAULT_OUTPUT
+from dagster.core.errors import DagsterError, DagsterInvariantViolationError
+from dagster.core.events import (
+    AssetObservationData,
+    DagsterEvent,
+    DagsterEventType,
+    StepMaterializationData,
+)
+from dagster.core.storage.pipeline_run import DagsterRun
+
+
+class ExecutionResult(ABC):
+    @property
+    def job_def(self) -> JobDefinition:
+        raise NotImplementedError()
+
+    @property
+    def dagster_run(self) -> DagsterRun:
+        raise NotImplementedError()
+
+    @property
+    def all_events(self) -> Sequence[DagsterEvent]:
+        raise NotImplementedError()
+
+    @property
+    def run_id(self) -> str:
+        raise NotImplementedError()
+
+    @property
+    def successful_steps(self) -> Sequence[str]:
+        raise NotImplementedError()
+
+    @property
+    def skipped_steps(self) -> Sequence[str]:
+        raise NotImplementedError()
+
+    @property
+    def failed_steps(self) -> Sequence[str]:
+        raise NotImplementedError()
+
+    @property
+    def success(self) -> bool:
+        """bool: Whether execution was successful."""
+        return self.dagster_run.is_success
+
+    @property
+    def all_node_events(self) -> Sequence[DagsterEvent]:
+        """List[DagsterEvent]: All dagster events from the in-process execution."""
+
+        step_events: List[DagsterEvent] = []
+
+        for node_name in self.job_def.graph.node_dict.keys():
+            handle = NodeHandle(node_name, None)
+            step_events += self._filter_events_by_handle(handle)
+
+        return step_events
+
+    @abstractmethod
+    def _get_output_for_handle(self, handle: NodeHandle, output_name: str) -> Any:
+        raise NotImplementedError()
+
+    def _filter_events_by_handle(self, handle: NodeHandle) -> Sequence[DagsterEvent]:
+        def _is_event_from_node(event: DagsterEvent) -> bool:
+            if not event.is_step_event:
+                return False
+            node_handle = cast(NodeHandle, event.solid_handle)
+            return node_handle.is_or_descends_from(handle)
+
+        return self.filter_events(_is_event_from_node)
+
+    def output_value(self, output_name: str = DEFAULT_OUTPUT) -> Any:
+        """Retrieves output of top-level job, if an output is returned.
+
+        If the top-level job has no output, calling this method will result in a
+        DagsterInvariantViolationError.
+
+        Args:
+            output_name (Optional[str]): The name of the output to retrieve. Defaults to `result`,
+                the default output name in dagster.
+
+        Returns:
+            Any: The value of the retrieved output.
```

Comment:
[1] the shared impl API docs don't telegraph having to open the context manager. Lil gross for the base class docs to have to know about / refer to its sub classes. Could override the method and just call the parent to provide different doc strings.